Mathlib/Analysis/Calculus/ContDiff/Basic.lean

```diff
@@ -907,6 +907,39 @@ theorem ContDiff.smulRight {f : E → F →L[𝕜] 𝕜} {g : E → G} {n : ℕ
 
 end SpecificBilinearMaps
 
+section ClmApplyConst
+
+/-- Application of a CLM to a constant commutes with `iteratedFDerivWithin`. -/
+theorem iteratedFDerivWithin_clm_apply_const_apply
+    {s : Set E} (hs : UniqueDiffOn 𝕜 s) {n : ℕ∞} {c : E → F →L[𝕜] G} (hc : ContDiffOn 𝕜 n c s)
+    {i : ℕ} (hi : i ≤ n) {x : E} (hx : x ∈ s) {u : F} {m : Fin i → E} :
+    (iteratedFDerivWithin 𝕜 i (fun y => (c y) u) s x) m = (iteratedFDerivWithin 𝕜 i c s x) m u := by
+  induction i generalizing x with
+  | zero => simp
+  | succ i ih =>
+    replace hi : i < n := lt_of_lt_of_le (by norm_cast; simp) hi
+    have h_deriv_apply : DifferentiableOn 𝕜 (iteratedFDerivWithin 𝕜 i (fun y => (c y) u) s) s :=
+      (hc.clm_apply contDiffOn_const).differentiableOn_iteratedFDerivWithin hi hs
+    have h_deriv : DifferentiableOn 𝕜 (iteratedFDerivWithin 𝕜 i (fun y => c y) s) s :=
+      hc.differentiableOn_iteratedFDerivWithin hi hs
+    simp [iteratedFDerivWithin_succ_apply_left]
+    rw [← fderivWithin_continuousMultilinear_apply_const_apply (hs x hx) (h_deriv_apply x hx)]
+    rw [fderivWithin_congr' (fun x hx => ih (le_of_lt hi) hx) hx]
+    rw [fderivWithin_clm_apply (hs x hx) (h_deriv.continuousMultilinear_apply_const _ x hx)
+      (differentiableWithinAt_const u)]
+    simp [fderivWithin_const_apply _ (hs x hx)]
```

Comment:
In mathlib, we'd like to avoid a bare `simp` call in the middle of a proof. `simp only` is fine (you can use `simp?` to find such an invocation), as is using `simp` when it closes a goal.

In mathlib jargon, these are called "non-terminal simps". They are harder to maintain over time (and using `simp only` can be slightly faster). The [glossary](https://leanprover-community.github.io/glossary.html#non-terminal-simp) (beware, written for Lean 3!) explains a bit more why they're avoided.

Reply:
Thanks! I didn't know that, and it makes sense